core/config/stop_config.cpp

```diff
@@ -0,0 +1,151 @@
+        if (auto& obj = config.get("reduction_factor")) {
+            factory.with_reduction_factor(
+                gko::config::get_value<remove_complex<ValueType>>(obj));
+        }
+        if (auto& obj = config.get("baseline")) {
+            factory.with_baseline(get_mode(obj.get_string()));
+        }
+        return factory;
+    }
+};
+
+
+inline deferred_factory_parameter<stop::CriterionFactory> configure_residual(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    auto updated = update_type(config, td);
+    return dispatch<stop::CriterionFactory, ResidualNormConfigurer>(
+        updated.get_value_typestr(), config, context, updated,
+        value_type_list());
+}
+
+
+template <typename ValueType>
+class ImplicitResidualNormConfigurer {
+public:
+    static deferred_factory_parameter<
+        typename stop::ImplicitResidualNorm<ValueType>::Factory>
+    build_from_config(const gko::config::pnode& config,
+                      const gko::config::registry& context,
+                      gko::config::type_descriptor td_for_child)
+    {
+        auto factory = stop::ImplicitResidualNorm<ValueType>::build();
+        if (auto& obj = config.get("reduction_factor")) {
+            factory.with_reduction_factor(
+                gko::config::get_value<remove_complex<ValueType>>(obj));
+        }
+        if (auto& obj = config.get("baseline")) {
+            factory.with_baseline(get_mode(obj.get_string()));
+        }
+        return factory;
+    }
+};
+
+
+inline deferred_factory_parameter<stop::CriterionFactory>
+configure_implicit_residual(const pnode& config, const registry& context,
+                            type_descriptor td)
+{
+    auto updated = update_type(config, td);
+    return dispatch<stop::CriterionFactory, ImplicitResidualNormConfigurer>(
+        updated.get_value_typestr(), config, context, updated,
+        value_type_list());
+}
+
+
+template <>
+deferred_factory_parameter<const stop::CriterionFactory>
+get_factory<const stop::CriterionFactory>(const pnode& config,
+                                          const registry& context,
+                                          type_descriptor td)
+{
+    deferred_factory_parameter<const stop::CriterionFactory> ptr;
+    if (config.get_tag() == pnode::tag_t::string) {
+        return context.search_data<stop::CriterionFactory>(config.get_string());
+    } else if (config.get_tag() == pnode::tag_t::map) {
+        static std::map<std::string,
+                        std::function<deferred_factory_parameter<
+                            gko::stop::CriterionFactory>(
+                            const pnode&, const registry&, type_descriptor)>>
+            criterion_map{
+                {{"Time", configure_time},
+                 {"Iteration", configure_iter},
+                 {"ResidualNorm", configure_residual},
+                 {"ImplicitResidualNorm", configure_implicit_residual}}};
```

Comment:
I think the criteria should be handles in a more direct way, compared to other ginkgo types, similar to what Tobias did.